airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/state/StreamStateManager.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.source.relationaldb.state;
+
+import static io.airbyte.integrations.source.relationaldb.state.StateGeneratorUtils.CURSOR_FIELD_FUNCTION;
+import static io.airbyte.integrations.source.relationaldb.state.StateGeneratorUtils.CURSOR_FUNCTION;
+import static io.airbyte.integrations.source.relationaldb.state.StateGeneratorUtils.NAME_NAMESPACE_PAIR_FUNCTION;
+
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.base.AirbyteStreamNameNamespacePair;
+import io.airbyte.integrations.source.relationaldb.CdcStateManager;
+import io.airbyte.integrations.source.relationaldb.CursorInfo;
+import io.airbyte.integrations.source.relationaldb.models.CdcState;
+import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage.AirbyteStateType;
+import io.airbyte.protocol.models.AirbyteStreamState;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Per-stream implementation of the {@link StateManager} interface.
+ *
+ * This implementation generates a state object for each stream detected in catalog/map of known
+ * streams to cursor information stored in this manager.
+ */
+public class StreamStateManager extends AbstractStateManager<AirbyteStateMessage, AirbyteStreamState> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamStateManager.class);
+
+  /**
+   * Constructs a new {@link StreamStateManager} that is seeded with the provided
+   * {@link AirbyteStateMessage}.
+   *
+   * @param airbyteStateMessages The initial state represented as a list of
+   *        {@link AirbyteStateMessage}s.
+   * @param catalog The {@link ConfiguredAirbyteCatalog} for the connector associated with this state
+   *        manager.
+   */
+  public StreamStateManager(final List<AirbyteStateMessage> airbyteStateMessages, final ConfiguredAirbyteCatalog catalog) {
+    super(catalog,
+        () -> airbyteStateMessages.stream().map(a -> a.getStream()).collect(Collectors.toList()),
+        CURSOR_FUNCTION,
+        CURSOR_FIELD_FUNCTION,
+        NAME_NAMESPACE_PAIR_FUNCTION);
+  }
+
+  @Override
+  public CdcStateManager getCdcStateManager() {
```

Comment:
should this return null? it shouldn't be possible to have cdc state for stream manager.

Reply:
@cgardens This is an attempt to avoid changing too much code as part of this.  There are other sources (MySQL and MSSQL) that call this method on the existing state manager to directly update the CDC state.  In order to introduce an interface for the state manager and not have to add a bunch of class checks in those other sources, I decided simply to add this method to maintain backwards compatibility.  I started with returning null, but realized that would mean that the code that interacts with the CDC state manager would now need null checks, expanding the change.  Instead, I opted for returning the CDC state manager with no internal state to avoid breaking things.  In a perfect world, this goes away and those other sources are updated to stop setting the CDC state directly.

Reply:
I agree. I think instead of null, we should throw an exception

Reply:
cool. makes sense. thanks.

Reply:
This has been changed to throw an `UnsupportedOperationException` in the per stream state manager.